doc/diagnostics/WritingDdrCode.md

```diff
@@ -0,0 +1,174 @@
+error, `NoSuchMethodError`.
+
+It is also possible that there is a need to use a field in a structure that was not always
+present. In that situation, the linkage error would be `NoClassDefFoundError` if the pointer
+classes are being generated dynamically.
+
+Because linkage errors are unchecked exceptions, it is easy forget to handle those errors,
+with unwelcome consequences at runtime.
+
+The solution, introduced by [#12676](https://github.com/eclipse-openj9/openj9/pull/12676),
+makes explicit the set of fields that may be accessed by hand-written DDR code. There are two
+kinds of such fields, depending on whether they have always been present or not (they were
+either added or removed at some point). Those fields are listed in `AuxFieldInfo29.dat` along
+with their types. The word `required` marks those that must have always existed, for example:
+
+```
+J9Object.clazz = required
+```
+
+while an optional field gives its assumed type:
+
+```
+J9ModronThreadLocalHeap.realHeapTop = U8*
+```
+
+Pointer classes generated in source form during the build are unaffected by this extra field
+information. Those generated in binary form (i.e. in `.class` files) are limited so they only
+include methods corresponding to fields mentioned in `AuxFieldInfo29.dat` to prevent inadvertent
+use of a field (that has not always been present) in hand-written code.
+
+The source pattern for optional fields was modified to throw a checked exception (`NoSuchFieldException`):
+
+```
+    // U8* realHeapTop
+    @com.ibm.j9ddr.GeneratedFieldAccessor(offsetFieldName="_realHeapTopOffset_", declaredType="U8*")
+    public U8Pointer realHeapTop() throws CorruptDataException, NoSuchFieldException {
+        try {
+            return U8Pointer.cast(getPointerAtOffset(J9ModronThreadLocalHeap._realHeapTopOffset_));
+        } catch (NoClassDefFoundError | NoSuchFieldError e) {
+            throw new NoSuchFieldException();
+        }
+    }
+
+    // U8* realHeapTop
+    public PointerPointer realHeapTopEA() throws CorruptDataException, NoSuchFieldException {
+        try {
+            return PointerPointer.cast(nonNullFieldEA(J9ModronThreadLocalHeap._realHeapTopOffset_));
+        } catch (NoClassDefFoundError | NoSuchFieldError e) {
+            throw new NoSuchFieldException();
+        }
+    }
+```
+
+Dynamically generated classes include either this when the field is present (in a 64-bit core):
+
+```
+    public U8Pointer realHeapTop() throws CorruptDataException, NoSuchFieldException {
+        return U8Pointer.cast(getPointerAtOffset(8));
+    }
+```
+
+or this when the field is absent:
+
+```
+    public U8Pointer realHeapTop() throws CorruptDataException, NoSuchFieldException {
+        throw new NoSuchFieldException();
+    }
+```
+
+Code that accesses optional fields must handle `NoSuchFieldException` because it is a checked
+exception.
+
+To start using a field in DDR code, an entry must be added to `AuxFieldInfo29.dat`. If the field
```

Comment:
I suggest these two paragraphs either be moved or copied to the top of this document to provide a TLDR section.

Reply:
Done.